plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/index/HudiSecondaryIndexSupport.java

```diff
@@ -30,86 +33,101 @@
 import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
 import java.util.stream.Collectors;
 
+import static io.trino.plugin.hudi.HudiSessionProperties.getSecondaryIndexWaitTimeout;
 import static io.trino.plugin.hudi.query.index.HudiRecordLevelIndexSupport.constructRecordKeys;
 import static io.trino.plugin.hudi.query.index.HudiRecordLevelIndexSupport.extractPredicatesForColumns;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
 
 public class HudiSecondaryIndexSupport
         extends HudiBaseIndexSupport
 {
     private static final Logger log = Logger.get(HudiSecondaryIndexSupport.class);
+    private final CompletableFuture<Optional<Set<String>>> relevantFileIdsFuture;
+    private final Duration secondaryIndexWaitTimeout;
+    private final long futureStartTimeMs;
 
-    public HudiSecondaryIndexSupport(SchemaTableName schemaTableName, Lazy<HoodieTableMetaClient> lazyMetaClient)
+    public HudiSecondaryIndexSupport(ConnectorSession session, SchemaTableName schemaTableName, Lazy<HoodieTableMetaClient> lazyMetaClient, Lazy<HoodieTableMetadata> lazyTableMetadata, TupleDomain<HiveColumnHandle> regularColumnPredicates)
     {
         super(log, schemaTableName, lazyMetaClient);
-    }
-
-    @Override
-    public Map<String, List<FileSlice>> lookupCandidateFilesInMetadataTable(
-            HoodieTableMetadata metadataTable,
-            Map<String, List<FileSlice>> inputFileSlices,
-            TupleDomain<String> regularColumnPredicates)
-    {
+        this.secondaryIndexWaitTimeout = getSecondaryIndexWaitTimeout(session);
+        TupleDomain<String> regularPredicatesTransformed = regularColumnPredicates.transformKeys(HiveColumnHandle::getName);
         if (regularColumnPredicates.isAll() || lazyMetaClient.get().getIndexMetadata().isEmpty()) {
             log.debug("Predicates cover all data, skipping secondary index lookup.");
-            return inputFileSlices;
+            this.relevantFileIdsFuture = CompletableFuture.completedFuture(Optional.empty());
```

Comment:
`lazyMetaClient.get()` triggers metadata instantiation and timeline load in the constructor.  Could this check be added to the async execution?